### PR TITLE
Change LokiBackend to store using nanosecond timestamp when log was received

### PR DESCRIPTION
**Pull request by codewithcheese**
Change-type: patch
Signed-off-by: Thomas Manning <thomasm@balena.io>

### CODE REVIEW

File: src/lib/utils.ts [hunk 1]
```diff
@@ -4,6 +4,19 @@ import * as ipaddr from 'ipaddr.js';
...
+const loadNs = process.hrtime();
+const loadMs = new Date().getTime();
+
+export function getNanoTimestamp() {
+	const diffNs = process.hrtime(loadNs);
+	const loadTs = BigInt(loadMs) * 1000000n;
+	const secondsDiff = BigInt(diffNs[0]) * 1000000000n;
+	const nanosDiff = BigInt(diffNs[1]);
+	return loadTs + secondsDiff + nanosDiff;
+}
```
Comment (Page-) [inline src/lib/utils.ts:18]: You can avoid a lot of repeated calculations here by switching to:
```suggestion
const loadNs = BigInt(Date.now()) * 1000000n - process.hrtime.bigint();

export function getNanoTimestamp() {
	return process.hrtime.bigint() + loadNs;
}
```

File: src/features/device-logs/lib/backends/loki.ts [hunk 2]
```diff
@@ -328,21 +323,18 @@ export class LokiBackend implements DeviceLogsBackend {
...
 			const timestamp = new Timestamp();
-			timestamp.setSeconds(seconds);
-			timestamp.setNanos(nanoseconds);
+			timestamp.setSeconds(Math.floor(Number(log.nanoTimestamp / 1000000000n)));
+			timestamp.setNanos(Number(log.nanoTimestamp % 1000000000n));
 			// store log line as JSON
-			const logJson = JSON.stringify(log);
+			const logJson = JSON.stringify(log, (key, value) =>
+				key === 'nanoTimestamp' ? undefined : value,
+			);
```
Comment (Page-) [inline src/features/device-logs/lib/backends/loki.ts:337]: Do you know how this performs vs
```ts
delete log.nanoTimestamp
const logJson = JSON.stringify(log);
```
by any chance?

**Review by Page- — COMMENTED** [review]
I'm assuming the VERSION doesn't need changing because it only affects metadata?

File: src/features/device-logs/lib/backends/loki.ts (hunk 2 above)
Comment (codewithcheese) [reply src/features/device-logs/lib/backends/loki.ts:337]: Here's a benchmark

```
const loadNs = BigInt(Date.now()) * 1000000n - process.hrtime.bigint();

function getNanoTimestamp() {
	return loadNs + process.hrtime.bigint();
}

const createLog = (extra = {}) => {
	return {
		isStdErr: true,
		isSystem: true,
		message: `a log line`,
		nanoTimestamp: getNanoTimestamp(),
		timestamp: Date.now(),
		createdAt: Date.now(),
		...extra,
	};
};

function genLogs(removeNanos = false) {
	const logs = []
	for (let i = 0; i < 1e6; i++) {
		const log = createLog()
		if (removeNanos) {
			// @ts-ignore
			delete log.nanoTimestamp
		}
		logs.push(log)
	}
	return logs
}

let logs = genLogs()

let startTime = Date.now()
for (const log of logs) {
	// @ts-ignore
	delete log.nanoTimestamp
	JSON.stringify(log)
}
console.log('Delete took', Date.now() - startTime);

logs = genLogs()
startTime = Date.now()
for (const log of logs) {
	JSON.stringify(log, (key, value) =>
		key === 'nanoTimestamp' ? undefined : value)
}
console.log('Replacer took', Date.now() - startTime);

logs = genLogs(true)
startTime = Date.now()
for (const log of logs) {
	JSON.stringify(log)
}
console.log('Noop took', Date.now() - startTime);
```

Surprisingly delete is slower 
```
Delete took 2131
Replacer took 1664
Noop took 1141
```

File: src/lib/utils.ts (hunk 1 above)
Comment (codewithcheese) [reply src/lib/utils.ts]: Nice! was not aware they added a `bigint()`, updated.

File: src/features/device-logs/lib/backends/loki.ts (hunk 2 above)
Comment (Page-) [reply src/features/device-logs/lib/backends/loki.ts:341]: Yeah it's probably due to how it interacts with the hidden classes and fast props stuff, in that case I'm fine as-is

**Comment by codewithcheese** [comment]
@balena-ci I self-certify!

**Pull request merged**